### PR TITLE
Feature/col 544 azure

### PR DESCRIPTION
**Pull request by ayang99**
additional cleanup, updated README's outlining info about Azure implementation

### CODE REVIEW

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java [hunk 1]
```diff
@@ -86,10 +90,11 @@ public ObjectSpecification getProgress(File stateDir, String objectId) throws IO
         part.setMd5(completedPart.getMd5());
       } else {
         // part is not complete - check if it has expired
-        if (PresignedUrlValidator.isUrlExpired(part.getUrl())) {
+        // if (PresignedUrlValidator.isUrlExpired(part.getUrl())) {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java]: Please remove

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/AzurePresignedUrlValidator.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      // This is the expected date format from Azure
+      val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssX");
+      val zdt = ZonedDateTime.parse(se, formatter);
+
+      return zdt.withZoneSameInstant(ZoneId.systemDefault()).toLocalDateTime();
+    } else {
+      log.error("Could not identify expected date parameters in request: {}", flattenMap(args));
+      // missing expected query arguments
+      throw new IllegalArgumentException(
+          String.format("Could not parse presigned URL - missing expected expiry date parameters"));
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/AzurePresignedUrlValidator.java]: No need for `String.format`

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/AzurePresignedUrlValidator.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      val se = args.get("se");
+
+      // This is the expected date format from Azure
+      val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssX");
+      val zdt = ZonedDateTime.parse(se, formatter);
+
+      return zdt.withZoneSameInstant(ZoneId.systemDefault()).toLocalDateTime();
+    } else {
+      log.error("Could not identify expected date parameters in request: {}", flattenMap(args));
+      // missing expected query arguments
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/AzurePresignedUrlValidator.java]: Caps

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/PresignedUrlValidator.java [hunk 4]
```diff
@@ -69,64 +65,35 @@ public static LocalDateTime getExpiry(URL url) {
...
     log.trace("Specified Expiry: %s%n", expiry);
 
     return expiry.isBefore(now);
   }
 
-  static Map<String, String> collectQuery(List<NameValuePair> args) {
+  Map<String, String> collectQuery(List<NameValuePair> args) {
     val result = new ImmutableMap.Builder<String, String>();
     for (val pair : args) {
       // store keys as lower case by convention
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/PresignedUrlValidator.java]: Caps

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/AzurePresignedUrlValidator.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+public class AzurePresignedUrlValidator extends PresignedUrlValidator {
+
+  /**
+   * Return expiry date of presigned URL in the local time zone (system default).
+   * @param args - URL query arguments
+   * @return expiry date translated to specified time zone
+   */
+  @Override
+  LocalDateTime extractExpiryDate(Map<String, String> args, ZoneId effectiveTimeZone) {
+    if (args.containsKey("se")) {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/AzurePresignedUrlValidator.java]: Can we save this to a constant and add comment about what it signifies?

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      String ts = "";
+      try {
+        // using string literals for query parameter names because the values found here:
+        // http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/constant-values.html
+        // don't appear consistent: all lower case, no X-Amz-Expired constant at all.
+
+        ts = args.get("X-Amz-Date".toLowerCase()); // basic format ISO 8601 string in UTC
+        // extended format uses hyphens. Named formatters in Java 8 (like ISO_INSTANT) are extended format only
+        // So, we have to define formatter with explicit ISO 8601 basic format
+        val formatter = DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmssX");
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java]: Constant

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      val explicit = Instant.ofEpochMilli(Long.parseLong(ts) * 1000);
+      expiry = LocalDateTime.ofInstant(explicit, effectiveTimeZone);
+    } else if (args.containsKey("X-Amz-Expires".toLowerCase()) && (args.containsKey("X-Amz-Date".toLowerCase()))) {
+      String ts = "";
+      try {
+        // using string literals for query parameter names because the values found here:
+        // http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/constant-values.html
+        // don't appear consistent: all lower case, no X-Amz-Expired constant at all.
+
+        ts = args.get("X-Amz-Date".toLowerCase()); // basic format ISO 8601 string in UTC
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java]: Constant.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  LocalDateTime extractExpiryDate(Map<String, String> args, ZoneId effectiveTimeZone) {
+    LocalDateTime expiry = null;
+    if (args.containsKey("Expires".toLowerCase())) {
+      val ts = args.get("Expires".toLowerCase());
+      val explicit = Instant.ofEpochMilli(Long.parseLong(ts) * 1000);
+      expiry = LocalDateTime.ofInstant(explicit, effectiveTimeZone);
+    } else if (args.containsKey("X-Amz-Expires".toLowerCase()) && (args.containsKey("X-Amz-Date".toLowerCase()))) {
+      String ts = "";
+      try {
+        // using string literals for query parameter names because the values found here:
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java]: Caps

File: dcc-storage-server/README.md [hunk 9]
```diff
@@ -16,11 +16,41 @@ The storage server is comprised of the following components:
...
+mvn -am -pl dcc-storage/dcc-storage-server
 ```
 
 ## Logging
 
 To enable logging of request bodies and headers, append the following to the command line:
 
 `--logging.level.org.apache.http=DEBUG`
+
+## Azure
```
Comment (btiernay) [inline dcc-storage-server/README.md]: Great, thanks for this!

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureURLGenerator.java [hunk 10]
```diff
@@ -44,7 +44,9 @@
 @Slf4j
 public class AzureURLGenerator implements URLGenerator {
 
-  private static DateTimeFormatter df = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
+  // Formatter just for local logging/display
+  private static DateTimeFormatter df = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ").withZone(
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureURLGenerator.java]: `final`

File: icgc-storage-client/README.md [hunk 11]
```diff
@@ -21,3 +21,28 @@ alias icgc-storage-client="docker run -it --rm  -u $(id -u):$(id -g) -e ACCESSTO
...
+## Azure
+
+Profile: ``azure``
+
+Experimental functionality 
+The Storage Client running the ``azure`` profile works the same way as ``aws`` and ``collab`` with the following exceptions:
+
+1. the number of parallel upload/download threads needs to be set very high (to offset the fact that blocks are so small). Settings like:
+
+	```
```
Comment (btiernay) [inline icgc-storage-client/README.md]: "```yaml"

File: icgc-storage-client/README.md [hunk 12]
```diff
@@ -21,3 +21,28 @@ alias icgc-storage-client="docker run -it --rm  -u $(id -u):$(id -g) -e ACCESSTO
 # Usage with an example object-id from https://dcc.icgc.org
 icgc-storage-client download --object-id 5b845b9a-3dcd-59ef-9f56-9a99396e988f --output-dir /data --output-layout bundle
 ```
+
+## Azure
```
Comment (btiernay) [inline icgc-storage-client/README.md]: Thanks!

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+   * @return expiry date translated to specified time zone
+   */
+  @Override
+  LocalDateTime extractExpiryDate(Map<String, String> args, ZoneId effectiveTimeZone) {
+    LocalDateTime expiry = null;
+    if (args.containsKey("Expires".toLowerCase())) {
+      val ts = args.get("Expires".toLowerCase());
+      val explicit = Instant.ofEpochMilli(Long.parseLong(ts) * 1000);
+      expiry = LocalDateTime.ofInstant(explicit, effectiveTimeZone);
+    } else if (args.containsKey("X-Amz-Expires".toLowerCase()) && (args.containsKey("X-Amz-Date".toLowerCase()))) {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java]: Helper method to encapsulate the meaning of this predicate would be nice.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+  /**
+   * Return expiry date of presigned URL in the local time zone (system default). Distinguished between V2 and V4 AWS
+   * signatures
+   * @param args - URL query arguments
+   * @return expiry date translated to specified time zone
+   */
+  @Override
+  LocalDateTime extractExpiryDate(Map<String, String> args, ZoneId effectiveTimeZone) {
+    LocalDateTime expiry = null;
+    if (args.containsKey("Expires".toLowerCase())) {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java]: Same here and constants.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java (hunk 13 above)
Comment (ayang99) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/util/S3PresignedUrlValidator.java]: Looked here originally: http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/constant-values.html here, but found it in an internal class

**Comment by btiernay** [comment]
Thanks!

**Comment by btiernay** [comment]
https://github.com/aws/aws-sdk-java/blob/8ca2018dbc4a7958c3626171e28ec467bd4a32a7/aws-java-sdk-core/src/main/java/com/amazonaws/auth/internal/SignerConstants.java#L34

**Pull request merged**